c2/cuda/groth16_srs.cuh

```diff
@@ -178,20 +177,33 @@ private:
                 } h, l, a, b_g1, b_g2;
             } data;
 
+            if (!ngpus()) {
+                throw sppark_error{CUDA_ERROR_NO_DEVICE,
+                                   std::string("No CUDA devices available")};
+            }
+
             int srs_file = open(srs_path, O_RDONLY);
 
-            // TODO, replace asserts with custom exceptions
-            assert(srs_file >= 0);
+            if (srs_file < 0) {
+                throw sppark_error{-errno, fmt_errno(errno, "open(\"%s\") failed: ",
+                                                     srs_path)};
```

Comment:
Hmm, again, as I see the code... It's not impossible for `fmt_errno` to reset the `errno`, which means that the order in which the arguments are evaluated is significant. It's probably easier to program around the ambiguity than to confirm if the order is defined in the standard. Easiest would probably be to have `fmt_errno` return a `std::pair` and add a constructor that accepts it...

Reply:
So that it would look as `throw sppark_error{fmt_errno(errno, ...)}`. Well, even this looks kind of repetitive... I suppose one can move the entirety of fmt_errno to the constructor...

Reply:
Sounds good. If we do that, I would prefer to also keep the current constructor's functionality where one passes just an integer and a string to the constructor.

Reply:
Done!

> Sorry about the inconvenience of having to redo everything :-)

No problem!